TypeScript (#1819)
enkidevs/curriculum: typescript/introduction/tsc-basics/tsc-interfaces-intro.md

---
author: kapnobatai136

type: normal

category: must-know

aspects:
  - introduction

---

# Interfaces Intro

---
## Content

> One of the most popular data types that you will is the object. We had a small run-in with the `object`, `Object`, and `{}` types, but never explained how you can pre-define an object.

Reviewer: ```suggestion
The main data type that you can use as a collection of values and behaviors is called an object. We had a small run-in with the `object`, `Object`, and `{}` types, but never explained how you can define an object.
```

```ts
function f(companyObj: { name: string }) {
  console.log(companyObj.name);
}
const someObject = {
  name: "Enki",
  product: "EnkiApp"
};
f(someObject);
```

> The function will only check that the `name` argument exist, and if it does it continues running.

Reviewer: ```suggestion
The function will verify that `companyObj` has at least one property called `name` that has to be of type `string`. It won't care about any additional properties.
```

> We can re-write this function using interfaces, to increase readability:

Reviewer: If we wanted to reuse the type of `companyObj` at multiple places in our program, we can extract it into an interface:

```ts
interface Company {
  name: string;
}

function f(companyObj: Company) {
  // ...
}
```

> The order of the fields doesn't matter, as long as TypeScript can find the field it needs it will keep running. This is just a short introduction to interfaces, with the following mission going into more details.

Reviewer: ```suggestion
The order of the fields doesn't matter, TypeScript verifies their type. This is just a short introduction to interfaces, with the following mission going into more details.
```

---
## Practice

Create an interface for the object:

```ts
const myObject = {
  name: 'foobar',
  isTrue: false
}

??? MyInterface {
  name: ???;
  isTrue: ???;
}
```

> * interface
> * string
> * boolean
> * 'foobar'
> * false
> * number

Reviewer: ```suggestion
* interface
* string
* boolean
* 'foobar'
* false
* number
* Interface
```
